[devbox global] Added init hook flag to global shellenv

mohsenari opened the pull request:
## Summary
Added `--init-hook` flag to devbox global shellenv

## How was it tested?
- `./devbox global add hello`
- edit global devbox.json `code ~/.local/share/devbox/global/default/devbox.json` and add init_hook
- `./devbox global shellenv --init-hook`
- confirm the line `. ~/.local/share/devbox/global/default/.devbox/gen/scripts/.hooks.sh` is included in the output.
- `cat ~/.local/share/devbox/global/default/.devbox/gen/scripts/.hooks.sh` and confirm the init_hook is included in the file.

savil commented on internal/boxcli/global.go:
```diff
@@ -166,6 +168,25 @@ func pullGlobalCmdFunc(cmd *cobra.Command, args []string) error {
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	output, err := box.PrintEnv(cmd.Context(), false, runInitHook)
```
I think Mike added the useCache flag because he needed it in the generated wrappers code. For global, how do wrappers work @mikeland86?

If we keep the `false`, could we add a comment for the param name for readability: like `, false /* paramName*/`?

savil commented on internal/impl/devbox.go:
```diff
@@ -269,6 +269,9 @@ func (d *Devbox) PrintEnv(ctx context.Context, useCache bool, includeHooks bool)
 			return "", err
 		}
 	}
+	if err := d.ensurePackagesAreInstalled(ctx, ensure); err != nil {
+		return "", err
+	}
```
Looking at the code here, I think we need to make a few changes:

1. Drop lines 267-271 that call `d.Generate` if `d.isDotDevboxVersionCurrent`. I think this is redundant by the `generateShellFiles` inside `ensurePackagesAreInstalled`.
2. Inside `ensurePackagesAreInstalled`, we currently early return if `lockfile.IsUpToDate`. I think we should also confirm `isDotDevboxVersionCurrent`. Reason: we may change some logic for file generation or package installation in a new devbox release.
3. With the above changes, I think we will not need `d.writeScriptsToFiles`.

savil replied:
On a similar motivation to point (2) above, I think we should also check `isDotDevboxVersionCurrent` in other locations that early return if `lockfile.IsUpToDate`. 

I think there is only one other location: `d.nixEnv`

mikeland86 replied:
Are we sure we want to run `ensurePackagesAreInstalled` here? There are tradeoffs. On the first run this will install all missing packages which may be slow.

savil commented on internal/impl/devbox.go:
```diff
@@ -284,6 +287,10 @@ func (d *Devbox) PrintEnv(ctx context.Context, useCache bool, includeHooks bool)
+		err = d.writeScriptsToFiles()
+		if err != nil {
+			return "", err
+		}
```
I'm not a fan of this change. 

Just like with unified-env, I think we should centralize how we generate the files so that we understand exactly when the files are written. 

This will let us control correctness, and can also optimize that codepath via caching-and-invalidation.

savil replied:
please see my suggestion above for how we could ensure the `writeScriptsToFiles` is called from within `ensurePackagesAreInstalled`.

savil reviewed (commented):
WDYT of the comment below about improving the cache-invalidation logic and centralizing where we generate files.

mikeland86 commented:
This addresses some of the simplifying I suggested in my comment: https://github.com/jetpack-io/devbox/pull/921

mohsenari commented:
Thanks @mikeland86 for making changes that simplified this PR a lot. 

Per your suggestions I removed `d.ensurePackagesAreInstalled` from PrintEnv() and put `d.generateShellFiles()` with a comment on why it's added. 

I also removed `impl.GenerateShellEnv()` and instead of checking if we are in a devbox shell or not when running `devbox shellenv`, I removed `DEVBOX_SHELL_ENABLED=1` from the output of shellenv. Which I believe is the correct thing to do since shellenv should not care if you're in a devbox shell or not. It should just compute and print the environment.

cc: @savil

mohsenari commented on internal/boxcli/global.go:
```diff
@@ -102,14 +100,18 @@ func globalPullCmd() *cobra.Command {
 		Use:   "shellenv",
 		Short: "Print shell commands that add global Devbox packages to your PATH",
-		Run: func(cmd *cobra.Command, _ []string) {
-			fmt.Fprintln(cmd.OutOrStdout(), impl.GenerateShellEnv())
```
removed

mikeland86 commented on internal/impl/devbox.go:
```diff
@@ -272,6 +272,12 @@ func (d *Devbox) PrintEnv(ctx context.Context, includeHooks bool) (string, error
 			return "", err
 		}
 	}
+	// this ensures shell files are generated in case if user runs
+	// shellenv without running devbox shell before or
+	// running devbox global shellenv --init-hook
+	if err := d.generateShellFiles(); err != nil {
+		return "", err
+	}
```
Yep, agree with @savil. If user modifies devbox.json in any way `lock.IsUpToDate` will be false.

mohsenari replied:
Fair, I removed it. Also added d.ensurePackagesAreInstalled() to PrintEnv()

savil commented on internal/impl/devbox.go:
```diff
@@ -285,6 +291,11 @@ func (d *Devbox) PrintEnv(ctx context.Context, includeHooks bool) (string, error
+	// This removes DEVBOX_SHELL_ENABLED=1 from the output. This is better than
+	// putting checks if user is already in a shell or if user is using devbox global.
+	// Overall, devbox shellenv should not set or modify DEVBOX_SHELL_ENABLED
+	envStr = strings.ReplaceAll(envStr, "export DEVBOX_SHELL_ENABLED=\"1\";", "")
```
thinking about this again, do we want to remove this env-var?

while I understand how it helps our implementations by not needing to check for it, I worry that we are straying from returning the true "environment" that devbox would use for the shell. This `envStr` is returned from `devbox shellenv` after all.

WDY all T?

mohsenari replied:
Moved this to shellrc file. Now only devbox shell will have it.

mikeland86 replied:
rm from here?

savil reviewed (approved):
nice!

mikeland86 commented on internal/impl/devbox.go:
```diff
@@ -223,6 +225,8 @@ func (d *Devbox) RunScript(cmdName string, cmdArgs []string) error {
 	if err != nil {
 		return err
 	}
+	// Used to determine whether we're inside a shell (e.g. to prevent shell inception)
+	env["DEVBOX_SHELL_ENABLED"] = "1"
```
nit, add comment that this is temporary because services needs it but should probably go away (because run != devbox shell)
